zkvm/src/predicate.rs

```diff
@@ -27,21 +29,38 @@ pub enum Predicate {
     /// constructing aggregated signature.
     Key(VerificationKey),
 
-    /// Representation of a predicate as commitment to a program and blinding factor.
-    Program(Program, Vec<u8>),
-
-    /// Disjunction of n predicates.
-    Or(PredicateDisjunction),
+    /// Taproot Merkle tree.
+    Tree(PredicateTree),
 }
 
 #[derive(Clone, Debug)]
-pub struct PredicateDisjunction {
-    preds: Vec<Predicate>,
-    // We precompute the disjunction predicate when composing it via `Predicate::or()`,
+pub struct PredicateTree {
+    pub progs: Vec<Program>,
+    pub key: VerificationKey,
+    blinding: [u8; 32],
+    // We precompute the tweaked aggregated signing key when composing it via `Predicate::tree()`,
     // so that we can keep `to_point`/`encode` methods non-failable across all types.
     precomputed_point: CompressedRistretto,
 }
 
+#[derive(Clone, Debug, Default)]
+pub struct CallProof {
+    // Signing key
+    pub signing_key: VerificationKey,
+
+    // Bit pattern indicating neighbor position
+    pub positions: u32,
+
+    // List of left-right neighbors, excluding the root and leaf hash
+    pub neighbors: Vec<MerkleNeighbor>,
+}
+
+// PRTODO: Rewrite MerkleItem for Program using BlindedProgram.
+pub enum BlindedProgram {
```

Comment:
The more apt name for this is `TaprootLeaf` because it's an enumeration of two kinds of leaves: `Program` and `BlindingFactor`.

Reply:
Or `PredicateLeaf`.